src/soc/ibm/power9/homer.c

```diff
@@ -885,6 +887,42 @@ static void istep_16_1(int this_core)
 	//     p9_stop_save_scom() and others
 }
 
+static void getPpeScanRings(struct xip_hw_header *hw)
+{
+	struct ring_query query = {
+		.kwd_name = "#G",
+		.ring_id = EC_GPTR,
+		.min_instance_id = 0x20,
+		.max_instance_id = 0x20,
+		.ring_class = RING_CLASS_GPTR_EX,
+	};
+	enum ring_status ring_status;
+
+	uint8_t *homer_bar = (void *)hw;
+
+	static uint8_t ppe[16 * KiB];
+	uint32_t ppe_size = sizeof(ppe);
+
+	static uint8_t buf1[MAX_RING_BUF_SIZE];
+	static uint8_t buf2[MAX_RING_BUF_SIZE];
+	static uint8_t buf3[MAX_RING_BUF_SIZE];
+
+	tor_access_ring((struct tor_hdr *)(homer_bar + hw->rings.offset + 68),
+			EC_TIME, PT_CME, 0, ppe, &ppe_size, GET_PPE_LEVEL_RINGS);
```

Comment:
```suggestion
	struct tor_hdr *tor;
	copy_section(&tor, &hw->rings, hw, dd, FIND);
	tor_access_ring(tor,
			EC_TIME, PT_CME, 0, ppe, &ppe_size, GET_PPE_LEVEL_RINGS);
```

Just FYI how to use `copy_section()`. It will choose TOR for proper `dd`, which should be passed as an argument to `getPpeScanRings()`.